File: Design Pattern/Structural/Bridge/message_formatter.ts
interface MessageSender {
    sendMessage(content: string): void;
}

class EmailSender implements MessageSender {
    sendMessage(content: string): void {
        console.log(`Email: ${content}`);
    }
}

class SmsSender implements MessageSender {
    sendMessage(content: string): void {
        console.log(`SMS: ${content}`);
    }
}

abstract class Message {
    // TODO: Add a protected field for the MessageSender reference
    // TODO: Add a protected field for the content (string)
    protected sender: MessageSender;
    protected content: string;
    constructor(sender: MessageSender, content: string) {
        // TODO: Store the sender and content
        this.sender = sender;
        this.content = content;
    }

    abstract send(): void;
}

class TextMessage extends Message {
    constructor(sender: MessageSender, content: string) {
        super(sender, content);
    }

    send(): void {
        // TODO: Delegate to this.sender.sendMessage() with the content as-is
        this.sender.sendMessage(this.content);
    }
}

class UrgentMessage extends Message {
    constructor(sender: MessageSender, content: string) {
        super(sender, content);
    }

    send(): void {
        // TODO: Delegate to this.sender.sendMessage() with "[URGENT] " prepended to the content
        this.sender.sendMessage("[URGENT] " + this.content);
    }
}

const email: MessageSender = new EmailSender();
const sms: MessageSender = new SmsSender();
const m1 = new TextMessage(email, "Hello there");
const m2 = new UrgentMessage(sms, "Server is down");
m1.send();
m2.send();

File: Design Pattern/Structural/Bridge/message_formatter.cpp
#include <iostream>
#include <string>
using namespace std;

class MessageSender {
public:
  virtual ~MessageSender() {}
  virtual void sendMessage(const string &content) = 0;
};

class EmailSender : public MessageSender {
public:
  void sendMessage(const string &content) override {
    cout << "Email: " << content << endl;
  }
};

class SmsSender : public MessageSender {
public:
  void sendMessage(const string &content) override {
    cout << "SMS: " << content << endl;
  }
};

class Message {
protected:
  // TODO: Add a MessageSender* field
  // TODO: Add a string field for the content
  MessageSender *sender;
  string content;

public:
  Message(MessageSender *sender, const string &content) {
    // TODO: Store the sender pointer and content
    this->sender = sender;
    this->content = content;
  }

  virtual ~Message() {}

  virtual void send() = 0;
};

class TextMessage : public Message {
public:
  TextMessage(MessageSender *sender, const string &content)
      : Message(sender, content) {}

  void send() override {
    // TODO: Delegate to sender->sendMessage() with the content as-is
    sender->sendMessage(content);
  }
};

class UrgentMessage : public Message {
public:
  UrgentMessage(MessageSender *sender, const string &content)
      : Message(sender, content) {}

  void send() override {
    // TODO: Delegate to sender->sendMessage() with "[URGENT] " prepended to the
    // content
    sender->sendMessage("[URGENT] " + content);
  }
};

int main() {
  EmailSender email;
  SmsSender sms;
  TextMessage m1(&email, "Hello there");
  UrgentMessage m2(&sms, "Server is down");
  m1.send();
  m2.send();
  return 0;
}
